README.md
<div align='center'>

# @macchiatojs/router

---

[![Build Status][travis-img]][travis-url]
[![Coverage Status][coverage-img]][coverage-url]
[![NPM version][npm-badge]][npm-url]
![Code Size][code-size-badge]
[![License][license-badge]][license-url]

</div>

<!-- ***************** -->

[travis-img]: https://travis-ci.com/macchiatojs/router.svg?branch=master
[travis-url]: https://travis-ci.com/macchiatojs/router
[coverage-img]: https://coveralls.io/repos/github/macchiatojs/router/badge.svg?branch=master
[coverage-url]: https://coveralls.io/github/macchiatojs/router?branch=master
[npm-badge]: https://img.shields.io/npm/v/@macchiatojs/router.svg?style=flat
[npm-url]: https://www.npmjs.com/package/@macchiatojs/router
[license-badge]: https://img.shields.io/badge/license-MIT-green.svg?style=flat
[license-url]: https://github.com/macchiatojs/router/blob/master/LICENSE
[code-size-badge]: https://img.shields.io/github/languages/code-size/macchiatojs/router
[pr-welcoming-badge]: https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat
[trouter]: https://github.com/lukeed/trouter
[trek-router]: https://github.com/trekjs/router

<!-- ***************** -->

### Expressive elegant modern amiable glamorous Macchiato.js Router ⚡ <small> (support also raw Node.js and Koa.js) </small>.

## `Features`

- 🦄 Based on top of [Trouter][trouter] and/or [Trek Router][trek-router].
- 🚀 Isomorphic to the moon.
- 💅🏻 Express-style routing (`_.get`, `_.post`, `_.put`, `_.patch`, `_.delete`, etc.)
- 🔥 Blaze and lightweight router.
- ⚖️ Tiny Bundle.
- 🪁 Named URL parameters.
- 🎯 Route middleware.
- 🥞 Support router layer middlewares.
- 📋 Responds to `OPTIONS` requests with allowed methods.
- ⛔️ Support for `405 Method Not Allowed`.
- ❌ Support for `501 Path Not Implemented`.
- 🧼 Support `trailing slash` and `fixed path` by automatic redirection.
- ✨ Asynchronous support (`async/await`).
- 🐱‍👤 Support `Koa.js` and all framework which have the same behave.
- 🐢 Raw Node.js (`http`) support.
- 🎉 TypeScript support.

## `Installation`

```bash
# npm
$ npm install @macchiatojs/router
# yarn
$ yarn add @macchiatojs/router
```

## `Usage`

This is a practical example of how to use.

```typescript
import Macchiato, { Request, Response } from "@macchiatojs/kernel";
import Router from "@macchiatojs/router";

const app = new Macchiato();
const router = new Router(); // use trouter
// >>> some benchs say that trek-router have better perf than trouter. <<< //
// const router = new Router({ trek: true }); // use trek-router

router.get("/hello", (request: Request, response: Response) => {
  response.body = "Hello World";
});

app.use(router.routes());

app.start(2222);
```

with raw Node.js

```typescript
import http, { IncomingMessage, ServerResponse } from "http";
import Router from "@macchiatojs/router";

const router = new Router({ raw: true });

router.get("/hello", (request: IncomingMessage, response: ServerResponse) => {
  response.statusCode = 200;
  response.write("Hello World !");
  response.end();
});

const server = http.createServer(router.rawRoutes());

server.listen(2222);
```

with Koa.js

```typescript
import Koa from "koa";
import Router from "@macchiatojs/router";

const app = new Koa();
const router = new Router<Koa.Middleware>({ expressify: false });

router.get("/hello", (ctx: Koa.BaseContext) => {
  ctx.body = "Hello World !";
});

app.use(router.routes());

app.listen(2222);
```

## `API`

### Note:

> We use `@macchiatojs/kernel` (needed only when use TypeScript and/or Macchiato.js), `parseurl` (needed only when use raw Node.js), `@types/koa` (needed only when use TypeScript) and `koa` (needed only when use Koa.js) as peerDependencies.

### new Router(options?)

Create a new router.

| Param                | Type      | Description                                                                                    |
| -------------------- | --------- | ---------------------------------------------------------------------------------------------- |
| [options]            | `Object`  |                                                                                                |
| [options.prefix]     | `String`  | prefix router paths                                                                            |
| [options.expressify] | `Boolean` | use `express/connect style` when is `true` and `koa style` when is `false` (default to `true`) |
| [options.raw]        | `Boolean` | use `raw Node.js server` when is `true` (default to `false`)                                   |
| [options.trek]       | `Boolean` | use `trek-router` when is `true` and `trouter` when is `false` (default to `false`)            |

### router.get|post|put|patch|delete|all(path, handler)

The http methods provide the routing functionality in `router`.

Method middleware and handlers follow usual raw Node.js and express middleware or koa middleware behavior, except they will only be called when the method and path match the request.

```js
// handle a GET / request.

// raw Node Style
router.get("/", (request, response) => {
  response.statusCode = 200;
  response.write("Hello World !");
  response.end();
});

// Express/Connect Style
router.get("/", (request, response) => {
  response.send(200, "Hello World !");
});

// Koa Style
router.get("/", (ctx) => {
  ctx.response.send(200, "Hello World !");
});
```

### router.prefix(prePath)

Route paths can be prefixed at the router level:

```js
// handle a GET /prePath/users request.

// raw Node Style
router.prefix("/prePath").get("/", (request, response) => {
  response.statusCode = 200;
  response.write("Hello World !");
  response.end();
});

// Express/Connect Style
router.prefix("/prePath").get("/users", (request, response) => {
  response.send(200, "Hello World !");
});

// Koa Style
router.prefix("/prePath").get("/users", (ctx) => {
  ctx.response.send(200, "Hello World !");
});
```

### router.route(path)

Lookup route with given path.

```js
// handle a GET /users request.

// raw Node Style
router.prefix("/users").get((request, response) => {
  response.statusCode = 200;
  response.write("Hello World !");
  response.end();
});

// Express/Connect Style
router.route("/users").get((request, response) => {
  response.send(200, "Hello World !");
});

// Koa Style
router.route("/users").get((ctx) => {
  ctx.response.send(200, "Hello World !");
});
```

### router.use(...middlewares)

Use given middleware(s). Currently, use middleware(s) for all paths of router isntance.

### router.routes()

Returns router middleware which handle a route matching the request for `Macchiato.js` and `Koa.js`.

### router.rawRoutes()

Returns router middleware which handle a route matching the request for `raw Node.js`.

## `Support`

If you have any problem or suggestion please open an issue.

## `Related`

- [koa-ismorphic-router](https://github.com/3imed-jaberi/koa-isomorphic-router)

#### License

---

[MIT](LICENSE) &copy; [Imed Jaberi](https://github.com/3imed-jaberi)


CHANGELOG.md
# CHANGE LOG


### v0.11.0 - unreleased

- drop support for Node.js version less than 16.
- update all deps.
- imporve the code quality by use the modren syntax.
- add engine factory method to load middlewares engines.

### v0.10.1 (18 Nov 2021)

- fix typo on README.md file.

### v0.10.0 (6 Nov 2021)

- Fix CI by upgrade to v0.19.0 of @macchiatojs/kernel.

### v0.9.0 (2 Nov 2021)

- fix and improve ts types (cache, middelwareStore, trek-router ...etc).
- update TS to v4.4.4.
- update our custom middlewares to latest version (v0.2.0).

### v0.8.0 (29 Aug 2021)

- fix and improve types with last kernel types.
- update deps.

### v0.7.0 (22 Aug 2021)

- fix and improve types.
- update deps.

### v0.6.0 (28 Jun 2021)

- add `trek-router`.
- update README.md.

### v0.5.0 (28 Jun 2021)

- fix handleRoutes method by make it as private.
- add export default for CommonJS.
- add support for Koa.js (js + ts).
- update README.md.

### v0.4.0 (26 Jun 2021)

- add `parseurl` as peerDep.
- add better support for raw Node.js.
- coverage 100%.

### v0.3.0 (25 Jun 2021)

- add `@macchiatojs/middleware` and `@macchiatojs/koaify-middleware` deps.
- remove `parseurl` dep.
- add `@macchiatojs/kernel` as peerDeps.
- add support Macchiato.js middlewares styles (express/koa).
- test it with koa and express styles.
- coverage 100%.

### v0.2.0 (24 Jun 2021)

- adapte with raw nodejs, express, connect and koa handler style.
- test it with koa.
- use # over private keyword.

### v0.1.0 (24 Jun 2021)

- initial release.


test/koa.spec.ts
import request from 'supertest'
import assert from 'assert'
import Koa from 'koa'
import Router from '../src'
import { Next } from '@macchiatojs/kernel'

const METHODS = ['GET', 'POST', 'PUT', 'PATCH', 'DELETE']

describe('macchiatojs-router with koa.js', () => {
  let app: Koa

  beforeEach(() => {
    app = new Koa()
  })

  describe('self', () => {
    it('should return a object', () => {
      assert.strictEqual(typeof new Router<Koa.Middleware>({ expressify: false }), 'object')
    })

    it('should not return private props/method', () => {
      assert.deepStrictEqual(
        Object.getOwnPropertyNames(Router.prototype),
        [
          'constructor', 'get', 'post', 'put', 'patch',
          'delete', 'all', 'prefix', 'route', 'use', 'rawRoutes', 'routes'
        ]
      )
    })
  })

  describe('http verbs/methods', () => {
    it('get method', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })

      router.get('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'get data' }
      })

      app.use(router.routes() as any)

      request(app.listen())
        .get('/test')
        .expect('Content-Type', /json/)
        .expect(200, /get data/, done)
    })

    it('post method', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })

      router.post('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'post data' }
      })

      app.use(router.routes() as any)

      request(app.listen())
        .post('/test')
        .expect('Content-Type', /json/)
        .expect(200, /post data/, done)
    })

    it('put method', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })

      router.put('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'put data' }
      })

      app.use(router.routes() as any)

      request(app.listen())
        .put('/test')
        .expect('Content-Type', /json/)
        .expect(200, /put data/, done)
    })

    it('patch method', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })

      router.patch('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'patch data' }
      })

      app.use(router.routes() as any)

      request(app.listen())
        .patch('/test')
        .expect('Content-Type', /json/)
        .expect(200, /patch data/, done)
    })

    it('delete method', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })

      router.delete('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'delete data' }
      })

      app.use(router.routes() as any)

      request(app.listen())
        .delete('/test')
        .expect('Content-Type', /json/)
        .expect(200, /delete data/, done)
    })

    describe('all method', () => {
      const app = new Koa()
      const router = new Router<Koa.Middleware>({ expressify: false })

      router.all('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'delete data' }
      })

      app.use(router.routes() as any)

      METHODS.forEach((method) => {
        it(method.toLowerCase(), (done) => {
          request(app.listen())[method.toLowerCase()]('/test').expect(200, done())
        })
      })
    })

    it('use route to handle methods (chain)', (done) => {
      (() => {
        const router = new Router<Koa.Middleware>({ expressify: false })
        const app = new Koa()

        router
          .route('/test')
          .get((ctx: Koa.BaseContext) => { ctx.body = { msg: 'get data' } })
          .post((ctx: Koa.BaseContext) => { ctx.body = { msg: 'post data' } })
  
        app.use(router.routes() as any)

        request(app.listen())
        .get('/test')
        .expect('Content-Type', /json/)
        .expect(/get data/)
        .expect(200)

        request(app.listen())
          .post('/test')
          .expect('Content-Type', /json/)
          .expect(/delete data/)
          .expect(200)

        done()
      })()
    })
  })

  describe('allow header field', () => {
    it('should allow header functions work', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })
      const app = new Koa()
      
      router.get('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'get data' }
      })

      ;(['post','put','delete'] as string[]).forEach(m => { router[m]('/test', () => void 0) })
  
      app.use(router.routes() as any)

      request(app.listen())
        .get('/test')
        .expect(200, done)
    })
  })

  describe('params', () => {
    it('should return a response with params', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })
      const app = new Koa()
      
      router.get('/test/:state', (ctx: Koa.Context) => {        
        ctx.body = { msg: 'get data with ' + ctx.request['params'].state + ' as params' }
      })
  
      app.use(router.routes() as any)

      request(app.listen())
        .get('/test/work')
        .expect(/get data with work as params/)
        .expect(200, done)
    })
  })

  describe('trailing slash and fixed path', () => {
    it('should normalize the path and redirect to correct one', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })
      const app = new Koa()
      
      router.get('//test/', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'get data' }
      })
  
      app.use(router.routes() as any)
    
      request(app.listen())
        .get('//test')
        .expect(301)
        .expect('Location', '/test')
        .end(done)
    })
  })

  describe('options request', () => {
    it('should responds with allowed methods', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })
      const app = new Koa()
      
      router.get('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'get data' }
      })
  
      app.use(router.routes() as any)
    
      request(app.listen())
        .options('/test')
        .expect('Allow', 'GET')
        .expect(204, done)
    })
  })

  describe('405 method not allowed', () => {
    it('should responds with method not allowed', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })
      const app = new Koa()
      
      router.get('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'get data' }
      })
  
      app.use(router.routes() as any)
    
      request(app.listen())
        .post('/test')
        .expect(405, done)
    })
  })

  describe('501 path not implemented', () => {
    it('should responds with path not implemented', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })
      const app = new Koa()
      
      router.get('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'get data' }
      })
  
      app.use(router.routes() as any)

      request(app.listen())
        .get('/not-imp-test')
        .expect(501, done)
    })
  })

  describe('cache', () => {
    it('should get request from the cache', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })
      const app = new Koa()
      
      router.get('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'get data' }
      })
  
      app.use(router.routes() as any)
      const server = app.listen()
      // no cached request.
      request(server)
        .get('/test')
        .expect('Content-Type', /json/)
        .expect(/get/)
        .expect(200, () => {
          // cached request.
          request(server)
            .get('/test')
            .expect('Content-Type', /json/)
            .expect(/get/)
            .expect(200, done)
        })
    })
  })

  describe('route method', () => {
    it('should route method work', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })
      const app = new Koa()
      
      router.get('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'get data' }
      })
  
      app.use(router.routes() as any)

      request(app.listen())
        .get('/test')
        .expect('Content-Type', /json/)
        .expect(/get/)
        .expect(200, done)
    })
  })

  describe('prefix as instance arg and method', () => {
    // bad arg passed ==> don't use the prefix.
    it('should prefix instance arg work', (done) => {
      const router = new Router({ expressify: false, prefix: '/preRoute' })
      const app = new Koa()
      
      router.get('/test', (ctx: Koa.BaseContext) => {
        ctx.body = { msg: 'get data' }
      })
  
      app.use(router.routes() as any)

      request(app.listen())
        .get('/preRoute/test')
        .expect('Content-Type', /json/)
        .expect(/get/)
        .expect(200, done)
    })

    it('should prefix method work', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })
      const app = new Koa()

      router
        .prefix('/preRoute')
        .get('/test', (ctx: Koa.BaseContext) => {
          ctx.body = { msg: 'get data' }
        })
  
      app.use(router.routes() as any)

      request(app.listen())
        .get('/preRoute/test')
        .expect('Content-Type', /json/)
        .expect(/get/)
        .expect(200, done)
    })
  })

  describe('use method', () => {
    it('should use method work with good arg', (done) => {
      const router = new Router<Koa.Middleware>({ expressify: false })
      const app = new Koa()

      router
        .use((ctx: any, next: Next) => {
          console.log('logger', ctx.status, ctx.url)
          next()
        })
        .get('/test', async (ctx: Koa.BaseContext) => {
          ctx.body = { msg: 'get data' }
          return
        })
  
      app.use(router.routes() as any)

      request(app.listen())
        .get('/test')
        .expect('Content-Type', /json/)
        .expect(/get/)
        .expect(200, done)
    })

    it('should use method throw with bad arg', () => {
      assert.throws(() => { 
        const router = new Router<Koa.Middleware>({ expressify: false })
        const app = new Koa()
  
        router
        .use('bad args' as any)
        .get('/test', (ctx: Koa.BaseContext) => {
          ctx.body = { msg: 'get data' }
        })
    
        app.use(router.routes() as any)
      })
    })
  })
})

src/index.ts
/*!
 * @macchiatojs/router
 *
 *
 * Copyright(c) 2021-2022 Imed Jaberi
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 */

import Trouter from 'trouter'
// TODO: add trek-router ts types
import TrekRouter from 'trek-router'
import Middleware from '@macchiatojs/middleware'
import KoaifyMiddleware from '@macchiatojs/koaify-middleware'
import parse from 'parseurl'
import hashlruCache from 'hashlru'
import type { Methods } from 'trouter'
import type { IncomingMessage, ServerResponse } from 'http'
import type {
  MiddlewareEngine,
  Context,
  Request,
  Response,
  KeyValueObject,
  MacchiatoHandler,
  ExpressStyleHandler,
} from '@macchiatojs/kernel'
import type {
  Request as KoaRequest,
  Response as KoaResponse,
  Context as KoaContext,
} from 'koa'

/**
 * @types
 */
type rawHandler = ExpressStyleHandler<IncomingMessage, ServerResponse>
type Handler = MacchiatoHandler | rawHandler
type NormalizedRoute<T = Handler> = [T, KeyValueObject<string>]
type Route<T = Handler> =
  | { params: Record<string, string>; handlers: T[] }
  | NormalizedRoute<T>
type AllowHeaderStore = { path: string; methods: (string | Methods[])[] }
type HlruCachAccess = {
  get: (key: string | number) => unknown
  set: (key: string | number, value: unknown) => void
}
type EmptyString = ''
type InstanceOfMiddlewareEngine = InstanceType<
  typeof Middleware | typeof KoaifyMiddleware
>
type MiddlewareEngineFactory = (
  isRaw: boolean,
  isExpressify: boolean
) => InstanceOfMiddlewareEngine
type MacchiatoRouterOptions = {
  raw?: boolean
  expressify?: boolean
  trek?: boolean
  prefix?: string
  // TODO: update this option to be required
  engine?: MiddlewareEngineFactory
}

const defaultMiddlewareEngineFactory: MiddlewareEngineFactory = (
  isRaw: boolean,
  isExpressify: boolean
) =>
  isRaw
    ? new Middleware<IncomingMessage, ServerResponse>()
    : isExpressify
    ? new Middleware<Request, Response>()
    : new KoaifyMiddleware<Context | KoaContext>()

const uniqueArray = <T>(array: T[]) => [...new Set(array)]

/**
 * Isomorphic Router for Macchiato.js.
 *
 * @api public
 */
class Router<THandler = Handler> {
  // init attributes.
  #raw: boolean
  #expressify: boolean
  #router: TrekRouter | Trouter<THandler>
  #METHODS: Methods[]
  #routePrefix: string
  #routePath?: string
  #middlewaresStore: MiddlewareEngine<THandler>
  #cache: HlruCachAccess
  #allowHeaderStore: AllowHeaderStore[]

  // boot the router.
  constructor(options?: MacchiatoRouterOptions) {
    // init attributes.
    this.#raw = options?.raw ?? false
    this.#expressify = options?.expressify ?? true
    this.#router = options?.trek ? new TrekRouter() : new Trouter<THandler>()
    this.#METHODS = ['GET', 'POST', 'PUT', 'PATCH', 'DELETE']
    this.#routePrefix = options?.prefix || '/'
    // TODO: remove defaultMiddlewareEngineFactory in the future and use only engine and replace the passed args with env vars.
    this.#middlewaresStore =
      options?.engine?.(this.#raw, this.#expressify) ||
      defaultMiddlewareEngineFactory(this.#raw, this.#expressify)
    this.#cache = hashlruCache(1000)
    this.#allowHeaderStore = []
  }

  // normalize the path by remove all trailing slash.
  #normalizePath(path: string): string {
    path = path.replace(/\/\/+/g, '/')
    if (path !== '/' && path.slice(-1) === '/') path = path.slice(0, -1)
    return path
  }

  // get allow header for specific path.
  #getAllowHeaderTuple(path: string): AllowHeaderStore {
    return this.#allowHeaderStore.find(
      (allow) => allow.path === path
    ) as AllowHeaderStore
  }

  // register route with specific method.
  #on(
    method: Methods | Methods[] | EmptyString,
    path: string | THandler,
    ...middlewares: THandler[]
  ): this {
    // handle the path arg when passed as middleware.
    if (typeof path !== 'string') {
      middlewares = [path, ...middlewares]
      path = this.#routePath as string
    }

    // normalize the path.
    path = this.#normalizePath(this.#routePrefix + path)

    // register path with method(s) to re-use as allow header filed.
    // allow header.
    const allow = this.#getAllowHeaderTuple(path)

    // stock to allow header store with unique val array.
    // --> prepare the next allowed header values.
    const nextAllowedHeader = {
      path,
      methods: !allow ? [method] : uniqueArray([...allow.methods, method]),
    }
    // --> merge all allowed header values.
    this.#allowHeaderStore = [
      ...new Map(
        Object.entries([...this.#allowHeaderStore, nextAllowedHeader])
      ).values(),
    ]

    // register to route to the trouter stack.
    if (Array.isArray(method)) method = ''

    this.#router.add(method as string, path, ...middlewares)

    // give access to other method after use the current one.
    return this
  }

  // register route with get method.
  public get(path: string | THandler, ...middlewares: THandler[]): this {
    return this.#on('GET', path, ...middlewares)
  }

  // register route with post method.
  public post(path: string | THandler, ...middlewares: THandler[]): this {
    return this.#on('POST', path, ...middlewares)
  }

  // register route with put method.
  public put(path: string | THandler, ...middlewares: THandler[]): this {
    return this.#on('PUT', path, ...middlewares)
  }

  // register route with patch method.
  public patch(path: string | THandler, ...middlewares: THandler[]): this {
    return this.#on('PATCH', path, ...middlewares)
  }

  // register route with delete method.
  public delete(path: string | THandler, ...middlewares: THandler[]): this {
    return this.#on('DELETE', path, ...middlewares)
  }

  // `router.all()` method >> register route with all methods.
  public all(path: string | THandler, ...middlewares: THandler[]): this {
    return this.#on(this.#METHODS, path, ...middlewares)
  }

  // add prefix to route path.
  public prefix(prefix: string): this {
    this.#routePrefix = prefix
    return this
  }

  // give access to write once the path of route.
  public route(path: string): this {
    // update the route-path.
    this.#routePath = path

    // give access to other method after use the current one.
    return this
  }

  // use given middleware, if and only if, a route is matched.
  public use(...middlewares: THandler[]): this {
    // check middlewares.
    if (middlewares.some((mw) => typeof mw !== 'function'))
      throw new TypeError('`.use()` accept only function(s)')

    // add the current middlewares to the store.
    this.#middlewaresStore.push(...middlewares)

    // give access to other method after use the current one.
    return this
  }

  // normalize route to be the same on trouter and trek-router.
  #routeFactory(route: Route<THandler>): NormalizedRoute<THandler> {
    // trek-router normalizer behave.
    if (Array.isArray(route)) {
      const [handler, routeParams] = route
      const params = {}

      // const arraytoEntries = <T = unknown>(array: T[]) => new Map(array)
      // parse the params if exist.
      if (Array.isArray(routeParams) && routeParams.length > 0) {
        // convert the routeParams to object params.
        const trekRouterParams = Object.fromEntries(
          new Map(routeParams.map(param => [param.name, param.value]))
        )
        // merge global params with trekRouterParams
        Object.assign(params, trekRouterParams)
      }

      return [handler, params]
    }

    // trouter normalizer behave.
    return [route.handlers[0], route.params]
  }

  // router middleware which handle a route matching the request.
  #handleRoutes(
    request: Request | KoaRequest | IncomingMessage,
    response: Response | KoaResponse | ServerResponse
  ) {
    return async (
      context: Context | KoaContext | null = null
    ): Promise<void> => {
      // orignal path.
      const originalPath = this.#raw
        ? parse(request as IncomingMessage).pathname
        : (request as Request).path

      // normalize the path.
      const path = this.#normalizePath(originalPath)

      // ignore favicon request.
      // src: https://github.com/3imed-jaberi/koa-no-favicon/blob/master/index.js
      // if (/\/favicon\.?(jpe?g|png|ico|gif)?$/i.test(originalPath)) { return }

      // have slashs ~ solve trailing slash.
      if (path !== originalPath) {
        if (this.#raw) {
          this.#sendResponse(301, path)(response as ServerResponse)
          return
        }

        (response as Response).status = 301
        ;(response as Response).redirect(
          `${path}${(request as Request).search}`
        )
        return
      }

      // init route matched var.
      let route

      // generate the cache key.
      const requestCacheKey = `${request.method}_${originalPath}`
      // get the route from the cache.
      route = this.#cache.get(requestCacheKey)

      // if the current request not cached.
      if (!route) {
        // find route inside the routes stack.
        route = this.#router.find(request.method as Methods, originalPath)
        // put the matched route inside the cache.
        this.#cache.set(requestCacheKey, route)
      }

      // extract the handler func and the params array.
      const [handler, routeParams] = this.#routeFactory(route)

      // check the handler func isn't defined.
      if (!handler || handler['length'] === 0) {
        // get methods exist on allow header.
        const allowHeaderFiled = this.#getAllowHeaderTuple(path)

        if (allowHeaderFiled) {
          // if `OPTIONS` request responds with allowed methods.
          if (request.method === 'OPTIONS') {
            if (this.#raw) {
              this.#sendResponse(204, '', {
                Allow: allowHeaderFiled.methods.join(', '),
              })(response as ServerResponse)
              return
            }

            (response as Response).status = 204
            ;(response as Response).set(
              'Allow',
              allowHeaderFiled.methods.join(', ')
            )
            ;(response as Response).body = ''
            return
          }

          // support 405 method not allowed.
          if (this.#raw) {
            this.#sendResponse(
              405,
              `'${request.method}' is not allowed in '${originalPath}'.`,
              { Allow: allowHeaderFiled.methods.join(', ') }
            )(response as ServerResponse)
            return
          }

          (response as Response).status = 405
          ;(response as Response).set(
            'Allow',
            allowHeaderFiled.methods.join(', ')
          )
          ;(
            response as Response
          ).body = `'${request.method}' is not allowed in '${originalPath}'.`
          return
        }

        // suport 501 path not implemented.
        if (this.#raw) {
          this.#sendResponse(501, `'${originalPath}' not implemented.`, {
            Allow: '',
          })(response as ServerResponse)
          return
        }

        (response as Response).status = 501
        ;(response as Response).set('Allow', '')
        ;(response as Response).body = `'${originalPath}' not implemented.`
        return
      }

      // put the params object inside the request.
      request['params'] = routeParams

      // add the handler to middlewares stored by the `use` method.
      this.#middlewaresStore.push(handler)

      // run the middleware.
      await this.#middlewaresStore.compose(
        ...(!this.#expressify ? [context] : [request, response])
      )
    }
  }

  #expressifyRoutes(request: Request, response: Response): Promise<void> {
    return this.#handleRoutes(request, response)()
  }

  #koaifyRoutes(context: Context | KoaContext): Promise<void> {
    return this.#handleRoutes(context.request, context.response)(context)
  }

  #sendResponse(
    status: number,
    content: string,
    headers?: KeyValueObject<string>
  ) {
    return (response: ServerResponse): void => {
      response.statusCode = status

      if (headers) {
        for (const key in headers) {
          response.setHeader(key, headers[key])
        }
      }

      if (status === 301) {
        response.writeHead(301, { Location: `${content}` })
      } else {
        response.write(content)
      }

      response.end()
      return
    }
  }

  rawRoutes(): rawHandler {
    if (!this.#raw) {
      throw new Error('`.rawRoutes` used only with raw Node.js server')
    }

    return (request: IncomingMessage, response: ServerResponse) => {
      this.#handleRoutes(request, response)()
    }
  }

  routes(): MacchiatoHandler {
    return this.#expressify
      ? (request: Request, response: Response) => {
          this.#expressifyRoutes(request, response)
        }
      : (ctx: Context | KoaContext) => {
          this.#koaifyRoutes(ctx)
        }
  }
}

/**
 * Expose `Router`.
 */
export default Router

// Support CommonJS
module.exports = Router
